File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-5.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int res;
    private void dfs(TreeNode node, int val) {
        if (node == null) {
            return;
        }
        if (node.val >= val) {
            res += 1;
        }
        dfs(node.left, Math.max(node.val, val));
        dfs(node.right, Math.max(node.val, val));

    }
    public int goodNodes(TreeNode root) {
        res = 0;
        dfs(root, -101);
        return res;
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        max_length = 0
        def calc(node):
            nonlocal max_length
            if not node:
                return 0
            left = calc(node.left)
            right = calc(node.right)
            max_length = max(max_length, left + right)
            return 1 + max(left, right)
        calc(root)
        return max_length
        